crates/services/src/seqlock.rs

```diff
@@ -0,0 +1,158 @@
+//! A simple implementation of a sequential lock.
+//! More details: <https://docs.kernel.org/locking/seqlock.html>
+
+use std::{
+    cell::UnsafeCell,
+    sync::atomic::{
+        AtomicU64,
+        Ordering,
+    },
+};
+
+/// A simple implementation of a sequential lock.
+/// some usage of unsafe, T must be Copy
+#[derive(Debug)]
+pub struct SeqLock<T> {
+    sequence: AtomicU64,
+    data: UnsafeCell<T>,
+}
+
+unsafe impl<T: Send> Sync for SeqLock<T> {}
+
+/// The writer handle for the `SeqLock`.
+/// Only one writer exists for a `SeqLock`.
+#[derive(Debug)]
+pub struct SeqLockWriter<T> {
+    lock: std::sync::Arc<SeqLock<T>>,
+}
+
+impl<T> SeqLockWriter<T> {
+    /// Modifies the data within the lock.
+    pub fn write<F>(&self, f: F)
+    where
+        F: FnOnce(&mut T),
+    {
+        let lock = &self.lock;
+
+        // Indicate that a write operation is starting.
+        lock.sequence.fetch_add(1, Ordering::Release);
+
+        // Modify the data.
+        unsafe {
+            f(&mut *lock.data.get());
+        }
+
+        // Indicate that the write operation has finished.
+        lock.sequence.fetch_add(1, Ordering::Release);
+    }
+}
+
+/// The reader handle for the `SeqLock`.
+/// Multiple readers can be created for a `SeqLock`.
+#[derive(Clone, Debug)]
+pub struct SeqLockReader<T> {
+    lock: std::sync::Arc<SeqLock<T>>,
+}
+
+impl<T> SeqLockReader<T> {
+    /// Reads the data within the lock.
+    pub fn read(&self) -> T
+    where
+        T: Copy,
+    {
+        let lock = &self.lock;
+
+        loop {
+            // check starting guard
+            let start = lock.sequence.load(Ordering::Acquire);
+
+            // if odd, write in progress
+            if start % 2 != 0 {
+                continue;
+            }
+
+            let data = unsafe { *lock.data.get() };
+
+            // check starting/ending guard
+            let end = lock.sequence.load(Ordering::Acquire);
+
+            // if value changed, retry
+            if start == end && start % 2 == 0 {
+                return data;
+            }
+        }
+    }
+}
+
+impl<T> SeqLock<T> {
```

Comment:
If we require that `T: Copy` here, compiler will enforce that for us.

Reply:
addressed in [6d626b5](https://github.com/FuelLabs/fuel-core/pull/2524/commits/6d626b582438a4923cbfa4a399a03ee22ac6ffee)